### PR TITLE
Create a new infrastructure for presentational hints.

### PR DESCRIPTION
**Pull request by Ms2ger**
[Review on Reviewable](https://reviewable.io/reviews/servo/servo/5857)

### CODE REVIEW

**Comment by Ms2ger** [comment]
r? @pcwalton 

Further work should reduce `synthesize_presentational_hints_for_legacy_attributes` to a loop over parsed attributes, I think.

File: components/layout/css/matching.rs [hunk 1]
```diff
@@ -289,6 +289,12 @@ impl StyleSharingCandidate {
             return false
         }
 
+        let mut matching_rules = vec![];
+        element.synthesize_presentational_hints_for_legacy_attributes(&mut matching_rules);
+        if !matching_rules.is_empty() {
+            return false;
+        }
```
Comment (pcwalton) [inline components/layout/css/matching.rs]: This is probably going to be too slow. `can_share_style_with` is called zillions of times. Can you avoid creating the vector?

**Pull request merged**